test/xbgpu/test_engine.py

```diff
@@ -967,12 +1012,28 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
         )
         # assert_allclose converts to float, which bloats memory usage.
         # To keep it manageable, compare a batch at a time.
-        for i in range(beam_results.shape[0]):
-            for j in range(beam_results.shape[1]):
-                np.testing.assert_allclose(beam_results[i, j], expected_beams[i, j], atol=1)
-
+        for i in range(expected_beams.shape[0]):
+            # NOTE: As per the explanation at the end of `_send_data`, we
+            # only verify up the the penultimate heap's data for each `beam_result`
```

Comment:
You fixed the code below, but "penultimate" is still only coincidentally true for the particular test parameters.